implement logic to compute node location information

clement26695 opened the pull request:
Try to add location information on CstNodes

- [x] implement basic logic
- [x] make this optionnal/toggle-able
- [x] add test(s)
- [ ] performance analysis
- [ ] updating documentation

Close #932

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/SAP/chevrotain?pullRequest=943) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/SAP/chevrotain?pullRequest=943) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/SAP/chevrotain?pullRequest=943) it.</sub>

clement26695 commented:
I just added a test, but some other should certainly be added.

I was not sure where it had to be added, and what exactly to test : please let me know what should be changed !

kristianmandrup commented:
Awesome! I'd love to see some progress on this and would like to help out, when I find the time. Cheers!

bd82 commented on packages/chevrotain/test/utils/matchers.ts:
```diff
@@ -6,10 +6,24 @@ export function setEquality(actual: any[], expected: any[]): void {
+    startOffset = 1,
+    startLine = 1,
+    startColumn = 1,
+    endOffset = undefined,
```
why are we using undefined as the default values for the offsets?

bd82 commented on packages/chevrotain/test/parse/cst_spec.ts:
```diff
@@ -461,6 +461,70 @@ context("CST", () => {
+        expect(tokenStructuredMatcher(cst.children.first[0].children.A[0], A))
+            .to.be.true
+
+        expect(cst.children.first[0].location.startOffset).to.equal(1)
```
you can compare the whole object instead of each property individually 

```javascript
expect(x).to.deep.equal({  /* ... */ })
```

bd82 commented:
> I just added a test, but some other should certainly be added.

You should check that the position information is collected for in-lined rules too.
- https://sap.github.io/chevrotain/docs/guide/concrete_syntax_tree.html#in-lined-rules

bd82 commented:
Have a look here for benchmarking:
- https://github.com/SAP/chevrotain/blob/master/packages/chevrotain/benchmark_web/README.md

But make sure to rebase your changes with master first as there have been some performance fixes recently.

bd82 commented on packages/chevrotain/test/utils/matchers.ts:
```diff
@@ -6,10 +6,24 @@ export function setEquality(actual: any[], expected: any[]): void {
+    startOffset = 1,
+    startLine = 1,
+    startColumn = 1,
+    endOffset = Number.MAX_VALUE,
```
Not sure this matters what value we use but why are we using MAX_VALUE here?

clement26695 replied:
Indeed, I just came back from vacations a couple of days ago, I have to remember exactly what/why I changed :sweat_smile:

bd82 commented on packages/chevrotain/src/parse/parser/traits/tree_builder.ts:
```diff
@@ -67,7 +70,10 @@ export class TreeBuilder {
-            children: {}
+            children: {},
+            location: {
+                startOffset: Number.MAX_VALUE
```
I think the default should be a none valid value.
e.g: NaN ?

bd82 commented on packages/chevrotain/src/parse/cst/cst.ts:
```diff
@@ -51,6 +53,64 @@ export function addNoneTerminalToCst(
+    setNodeLocation(node, ruleResult.location)
+}
+
+function setNodeLocation(node: CstNode, locationInformation: any) {
```
1. maybe too many IF conditions here to check the properties exist, after the first tume we assign location info all the properties should exist, so maybe we can only check if one property is missing?

2. Do not allow casting to boolean in the conditions,
 - ```javascript
       // implicit casting
       if (x) {}
       // equivalent to:
        if (x !== undefined || if x !== [] || if x !== "" || x !== 0 /* ... */ ) {}
       // no casting and faster performance veersion
       if  (x !== undefined) {}
   ``

clement26695 replied:
Ok for 2

For 1, I was wondering: is all information (endOffset, startLine, endLine, etc...) always present ? If no, should we always return some value (like NaN) if not present ?

clement26695 commented:
Regarding performance, I got these numbers, with the current implementation (with rebasing with master) : 

![Capture du 2019-05-03 16-19-01](https://user-images.githubusercontent.com/25530077/57145105-c8b68a80-6dc2-11e9-9335-ca672ca62eca.png)

bd82 commented:
I assume you enabled the CST creation in the benchmark options:
- https://github.com/SAP/chevrotain/blob/master/packages/chevrotain/benchmark_web/parsers/options.js

What computer are you using? the overall performance seems pretty low, or perhaps you are not using Chrome to test?

clement26695 commented:
I used firefox as it didn't seem working on Chrome (some permissions issue, I guess: I got the error `Blocked a frame with origin "null" from accessing a cross-origin frame`)

bd82 commented:
V8 performance is the most important and chrome duplicates this.
I am not familiar with the problem you are describing as I never encountered it on multiple machines and OS.

Perhaps you can disable some web security policies during local development in your chrome.
- https://stackoverflow.com/questions/3102819/disable-same-origin-policy-in-chrome

clement26695 commented:
Now I have these results (with outputCst enabled in dev and last) : 
![Capture du 2019-05-03 17-31-59](https://user-images.githubusercontent.com/25530077/57148210-68771700-6dc9-11e9-87b1-4ff8d18e89d6.png)

compare to these with the master version:
![Capture du 2019-05-03 17-16-20](https://user-images.githubusercontent.com/25530077/57148243-788ef680-6dc9-11e9-809c-357d2386d87e.png)

So that's not really good :sweat_smile: 

I'll try to optimize with your suggestions

kristianmandrup commented:
Really awesome to see so much progress already. 

What is the current performance in "real numbers" (ms) for a typical use case, say for a typical file of around 1000 statements/expressions total? Any idea?
Needs to be less than 10ms for sure.

bd82 commented:
> What  is the current performance in "real numbers" (ms) for a typical use  case, say for a typical code file of around ~1000 nodes in total? Any  idea? Needs to be less than 10ms for sure.

That is a very wide question.
The inputs for the JSON/CSS benchmark are ~1,000 lines files so for the JSON use case you are looking at 1.2Milion lines of code per second (divide by 100 for 10ms).

The JavaScript code sample is around ~3,000 LOC however which makes the math less straightforward.

But it also depends on the grammar, the more complex the grammar the slower the parsing would be.
Some extreme edge cases it may be possible to get as low as 20-30K LOC/second (Java Parser implemented in the JHipster org).  So we are looking at a very wide range of 1 million to several tens of thousands LOC per second in general.

The problem is that the current implementation seems to reduce performance by 30-40% which is quite significant.

bd82 commented:
@clement26695  why did you post two pictures? the "dev" variant is with your changes and the "latest" is the latest released version on npm which should be very close to master. basically the benchmark is meant to compare upcoming version vs previous version to detect regressions.

clement26695 commented:
I didn't know if the performance fixes you mentioned were already released or not, so that's why I tested with the master branch (and as I saw the results, I saw some performance improvements) 

> But make sure to rebase your changes with master first as there have been some performance fixes recently.

clement26695 commented on packages/chevrotain/src/parse/cst/cst.ts:
```diff
@@ -51,6 +53,52 @@ export function addNoneTerminalToCst(
+        isNaN(node.location.startOffset) ||
+        node.location.startOffset > locationInformation.startOffset
+    ) {
+        node.location.startOffset = locationInformation.startOffset
```
I'm not sure about these : as the default value is NaN, we have to test if the current value is NaN before doing the comparison.

Other question : is the information (startOffset, endLine, etc) always present ? It could create some issues if it is missing

bd82 replied:
No its not always present:
https://sap.github.io/chevrotain/documentation/4_3_3/interfaces/ilexerconfig.html#positiontracking

Different levels of position tracking are possible which makes it trickier.
And worst yet we only know this once an input has been provided to the Parser.

In general I would create **four** different CSTPositionAgument methods and assign the relevant
one on the instance during constructor time, so the "correct" one would be in-lined for performance reasons and we would avoid conditional checks on properties that cannot exist.

but in this case we don't even know at constructor time what level of position tracking is needed
unless we provide such an option on the Parser constructor, and we probably should do so anyhow so maybe that is fine.

clement26695 replied:
That is what I was thinking... 

Why four CSTPositionAgument methods ? isn't it one for each positionTracking, so 3 ?

clement26695 replied:
Oh, I see, you mean 1 for each positionTracking option, and disabling any node location computation ?

bd82 replied:
Yep

kristianmandrup commented:
In general, I'd recommend to always to avoid (at least complex) `if` statements or even multi line functions and use extract method to encapsulate every piece of distinct logic.

PS: On a personal note I never use any statements at this point, such as `if`, `for`, `case`, ... Expressions are always a better option (more composable etc) 

You could simply do `x() || y(a,b) || z(a)` instead of using `if/else` or `case`.
Beautiful code :)

bd82 commented:
> PS: On a personal note I never use any statements at this point, such as if, for, case, ... Expressions are always a better option (more composable etc)

This may be related but we are discussing a different kind of optimization:

```javascript
let foo = function() { 
   console.log("hello")
}

if (process.env["flag"]) {
   // noop function...
   foo = function() {}
}

function bar() {
  // if the flag is enabled foo === NOOP
  // So once bar would be optimized (in-lined) by V8, this call would completely disappear 
  // and the optimized "bar" would only contain a single statement and thus be faster.
  foo() 
  console.log("world")
}



```

clement26695 commented:
I tried to implement it as you suggested.

I put 
```js
location: {
  startOffset: Number.MAX_VALUE,
  startLine: Number.MAX_VALUE,
  startColumn: Number.MAX_VALUE,
  endOffset: -1,
  endLine: -1,
  endColumn: -1
}
```
as default values, so that I don't have to check if the current values are NaN in the setNodeLocationXXX functions (=> check for NaN is commented right now, I'll remove it if your ok with that, ok put it again if you prefer using NaN as default values)

clement26695 commented:
Regarding benchmark, I have to admit that I'm not completely sure about the figures I have, as they can change radically between two runs, without making any changes in the code. I got these results :

<img width="797" alt="Screen Shot 2019-05-15 at 15 13 56" src="https://user-images.githubusercontent.com/25530077/57778455-647dba00-7724-11e9-890a-453329d64887.png">

clement26695 commented:
@bd82, did you had time to check this ? :smile:

bd82 commented:
> did you had time to check this ? 😄

Sorry nope, pretty busy lately... 😢

 but thanks for reminding me 👍
Hopefully I will have a look on the weekend.

bd82 commented:
O.k. finally got around to review this.
See for comments:
- https://github.com/SAP/chevrotain/commit/114a3b70a4ac1e453f1c873bc92536b8a60e0d0a

I think:
- In general this looks good.
- We can still likely improve the performance a bit.
- I am getting 20-30%% performance regression in the parser only test when doing full tracking.
  - This means 10-15% in the combined le.xer +parser flow.
  -  It is not great but it may be o.k. considering:
     - Not everyone needs full location tracking on CST Nodes.
     - We can likely improve the performance a bit (See comments in review).
     - Perhaps this option should not be enabled by default.

kristianmandrup commented:
Without knowing the details, I think for many typical use cases, the performance is ok at this point.
Let's get a solution that works "out in the open" and then we will see... Good job!
Cheers!

bd82 commented:
> Without knowing the details, I think for many typical use cases, the performance is ok at this point.
Let's get a solution that works "out in the open" and then we will see... Good job!
Cheers!

The performance is o.k. for the general case, but Chevrotain is not so fast without continued obsession with performance 😄 .

Also @clement26695 already detected a possible bug here:
- https://github.com/SAP/chevrotain/commit/114a3b70a4ac1e453f1c873bc92536b8a60e0d0a#r33667766

If you need this  feature **now** you can just add the relevant logic from this PR (with slight mods) to your parser by overriding   certain methods on your parser.
You can also add a post processing phase to your CST Nodes without  any changes to your parser.

But the official feature would have to be **fully ready** before it is merged.
- Although it seems already 80% done to me. (Great work @clement26695 ).

clement26695 commented:
I updated the logic with your suggestion and put a default value. 

I also added some tests for the different cases (full, onlyStart, onlyOffset, none, and undefined)

I think I just didn't edit the documentation yet

bd82 commented on packages/chevrotain/api.d.ts:
```diff
@@ -1624,6 +1624,16 @@ export interface CstNode {
+}
+
+export interface CstNodeLocation {
+    startOffset?: number
```
does not this always have at least the offset parts?

clement26695 replied:
It is always present for the tokens, but we can disable its computation for the nodes to optimize the performance, I think

bd82 replied:
What are our location collection modes?

1. Start + end offsets
2. Start + end line + column (I don't think it would make sense to collect lines without columns or vice versa)
3. Both 1 + 2.

We also simplify to two options.
1 offsets (End/start)
2. full.

Anyhow in that case we can define two/three interfaces for the CSTNodeLocation
each with **mandatory** (not optional properties) which will also help specify the types in the implementation.

```typescript
// e.g:
interface CSTNode {
  location: offsetsLoc || FullLoc
}
```

It is still not the easiest to use because how would an end user know which of he two/three union types
is relevant (they would have to do casting). but right now they also do not know which properties are really available... 

Do you have a better idea how to define this?

bd82 commented on packages/chevrotain/src/parse/parser/parser.ts:
```diff
@@ -65,7 +71,8 @@ export const DEFAULT_PARSER_CONFIG: IParserConfig = Object.freeze({
     errorMessageProvider: defaultParserErrorProvider,
-    serializedGrammar: null
+    serializedGrammar: null,
+    nodePositionTracking: <"full" | "onlyStart" | "onlyOffset" | "none">"none"
```
why do we need a cast here?

clement26695 replied:
if I don't cast, the error Type string is not assignable to type '"none" | "full ...' is thrown 

(sorry for the formatting, I'm on my phone...)

bd82 commented on packages/chevrotain/src/parse/parser/traits/tree_builder.ts:
```diff
@@ -19,13 +31,22 @@ export class TreeBuilder {
     LAST_EXPLICIT_RULE_STACK: number[]
 
+    // TODO: this method should have a better signature
+    setNodeLocation: Function
```
better signature needed...

bd82 commented on packages/chevrotain/src/parse/parser/traits/tree_builder.ts:
```diff
@@ -19,13 +31,22 @@ export class TreeBuilder {
+    // TODO: this method should have a better signature
+    setNodeLocation: Function
+    setInitialNodeLocation: Function
+    nodePositionTracking: "full" | "onlyStart" | "onlyOffset" | "none"
```
use the signature from the api.d.ts no need to duplicate.

bd82 commented on packages/chevrotain/src/parse/parser/traits/tree_builder.ts:
```diff
@@ -41,6 +62,62 @@ export class TreeBuilder {
             this.atLeastOneSepFirstInternal = this.atLeastOneSepFirstInternalNoCst
         }
+
+        // TODO: case insensitive options are an awesome idea.
```
remove my mistaken comment

bd82 commented on packages/chevrotain/src/parse/parser/traits/tree_builder.ts:
```diff
@@ -41,6 +62,62 @@ export class TreeBuilder {
+
+    setInitialNodeLocationOnlyOffset(cstNode: CstNode): void {
+        cstNode.location = {
+            startOffset: Infinity
```
Maybe the behavior should be different here than the Lexer.
In the lexer its easy to calculate endOffset if we use "onlyOffset" because we just add the token's length.

But in the CSTNode it won't be easy to calculate the endOffset from the startOffset.

bd82 replied:
We should probably give different names than the lexer modes to avoid confusion

clement26695 replied:
So you want 3 options, which can be combined?

- One with start/end info on offset
- One with start/end info on line
- One with start/end info on column

Also, I'm not sure on how to implement it : for instance, if we only have the startOffset info, should we compute the endOffset by getting the last token startOffset and adding its length ? It would be more difficult to compute the endColumn and endLine, if we don't have this info in the tokens

bd82 replied:
Lets simplify things so we can finish this feature 😄 

1. Lets have only two options
   - Offsets (start + end)
   - Full info

2. Lets document an assumption that if you want Full info you need to enable full location tracking in the lexer too.

clement26695 replied:
LGTM :smile:

bd82 commented on packages/chevrotain/src/parse/parser/parser.ts:
```diff
@@ -72,7 +72,7 @@ export const DEFAULT_PARSER_CONFIG: IParserConfig = Object.freeze({
     errorMessageProvider: defaultParserErrorProvider,
     serializedGrammar: null,
-    nodePositionTracking: <"full" | "onlyStart" | "onlyOffset" | "none">"none"
+    nodePositionTracking: <IParserConfig["nodePositionTracking"]>"none"
```
This is better but I meant something like this:
https://github.com/SAP/chevrotain/blob/master/packages/chevrotain/src/api.ts#L144

You can reference a type from the api.d.ts (type of x.y.z)

clement26695 replied:
I got this error, as IParserConfig is an interface, and not a type (I guess, I'm not a TypeScript expert)

![Capture d’écran de 2019-05-27 10-31-12](https://user-images.githubusercontent.com/25530077/58406780-a8e55000-806a-11e9-826a-4620573f554b.png)

bd82 replied:
you can define a custom type in api.d.ts

```typescript
export type NodePositionTrackingOptioins = "A" | "B" | "C" and reuse that type 
else where in the d.ts file or the .ts files
```

bd82 commented on packages/chevrotain/src/parse/parser/traits/tree_builder.ts:
```diff
@@ -32,9 +32,12 @@ export class TreeBuilder {
+        locationInformation: any
+    ) => void
+    setInitialNodeLocation: (cstNode: CstNode) => void
+    nodePositionTracking: IParserConfig["nodePositionTracking"]
```
I think you can reference the api.d.ts here too

bd82 commented on packages/chevrotain/api.d.ts:
```diff
@@ -1673,6 +1683,10 @@ export interface IParserConfig {
+    /**
+     * Enable computation of CST nodes location
+     */
+    nodePositionTracking?: "full" | "onlyStart" | "onlyOffset" | "none"
```
as mentioned elsewhere there is no sense is collecting "onlyStart" location when dealing with CSTNodes
which are often multi lines.

We should always collect pairs (end+start) of the location properties in CSTNode tracking

bd82 commented on packages/chevrotain/src/parse/cst/cst.ts:
```diff
@@ -24,11 +24,45 @@ import { GAstVisitor } from "../grammar/gast/gast_visitor_public"
+    nodeLocation: CstNodeLocation,
+    locationInformation: any
+): void {
+    if (nodeLocation.startOffset > locationInformation.startOffset) {
```
we are parsing left to right. so if we ignore error recovery scenarios
this condition can only be entered **once** (for the first token/subrule) of the CSTNode.
Maybe we should leave this for a future optimization though because
this adds complexity with handling error recovery Tokens.

bd82 commented on packages/chevrotain/src/parse/cst/cst.ts:
```diff
@@ -24,11 +24,45 @@ import { GAstVisitor } from "../grammar/gast/gast_visitor_public"
+): void {
+    setNodeLocationOnlyStart(nodeLocation, locationInformation)
+
+    if (nodeLocation.endOffset < locationInformation.endOffset) {
```
As mentioned before we are parsing left to right.
In this case it means we will always enter this condition (ignoring error recovery situations).

bd82 commented:
I am merging this to continue work on master branch.

Thanks @clement26695 for this awesome work! 👍
